java/code/src/com/redhat/rhn/manager/contentmgmt/ContentManager.java

```diff
@@ -195,16 +195,21 @@ public ContentProject updateProject(String label, Optional<String> newName, Opti
      * Remove Content Project
      *
      * @param label the label
+     * @param cleanTargets true if the linked environment targets (e.g. sw channels) should be removed
      * @param user the user
      * @throws PermissionException if given user does not have required role
      * @throws EntityNotExistsException if Content Project with given label is not found
      * @return the number of objects affected
      */
-    public int removeProject(String label, User user) {
+    public int removeProject(String label, boolean cleanTargets, User user) {
         ensureOrgAdmin(user);
-        return lookupProject(label, user)
-                .map(cp -> ContentProjectFactory.remove(cp))
-                .orElseThrow(() -> new EntityNotExistsException(label));
+        ContentProject project = lookupProject(label, user).orElseThrow(() -> new EntityNotExistsException(label));
+
+        if (cleanTargets) {
+            listProjectEnvironments(label, user).forEach(e -> removeEnvironment(e.getLabel(), label, user));
```

Comment:
Hmm, but `removeEnvironment()` seems to remove only the environment and not the software channels which are part of the env.
At least I cannot find code doing it.

Reply:
Additionally what should happen with clients using the channels which should be removed? 
We have other code which handle it. In that case a question come up to unsubscribe all clients from the channel which should be removed.

Reply:
> In that case a question come up to unsubscribe all clients from the channel which should be removed.

If we do this, it should be mentioned in the popup dialog.

Reply:
I think my question is already under discussion here so ---> https://github.com/uyuni-project/uyuni/pull/3987#issuecomment-880660952

I completely agree we should at least tell the user there are clients subscribed to channels (even better to have a link to point to a selection of them and list them in a proper view). Telling how many is an option but should not be a big deal once we know the info. For sure we should enhance the current option to be checked and warn the user about it within the same moment and context.

Reply:
I would suggest:

Check if there are clients using the channels. Depending on the result I would add one additional checkbox:
- If no clients are using the channels, provide the option to remove the channels as well
- If there are clients, tell the number of clients with the option to unsubscribe the clients and remove channels. The number can be a link pointing to a system list page with all the clients which are affected. E.g. ` There are 20 Clients still using the Channels (Targets or Environments?). Unsubscribe all clients from this project before removing it?`

I would not support the option "unsubscribe but keep the channels". I think this would be too much.

Reply:
Ok, the discussion got pretty hot here. Good :smile: 

@mcalmer removing an environment already removes channels as of today (it's done in the `ContentProjectFactory.purgeTarget`). I like your suggestion and I'll improve it that way. I'll also improve the screen for removing environments.

Thanks for the discussion!